x-pack/apm-server/profiling/ilm.go

```diff
@@ -0,0 +1,474 @@
+	"github.com/elastic/elastic-agent-libs/logp"
+	"github.com/elastic/elastic-agent-libs/monitoring"
+	es "github.com/elastic/go-elasticsearch/v8"
+	"github.com/elastic/go-elasticsearch/v8/esapi"
+)
+
+var (
+	ilmRegistry             = monitoring.Default.NewRegistry("apm-server.profiling.ilm")
+	rolloverCounter         = monitoring.NewInt(ilmRegistry, "custom_ilm.execution.count")
+	rolloverSkippedCounter  = monitoring.NewInt(ilmRegistry, "custom_ilm.skipped_for_time_constraints.count")
+	rolloverFailuresCounter = monitoring.NewInt(ilmRegistry, "custom_ilm.failed.count")
+	rolloverUndeletedIndex  = monitoring.NewInt(ilmRegistry, "custom_ilm.undeleted_index.count")
+	keyValueIndices         = []string{StackTraceIndex, StackFrameIndex, ExecutablesIndex}
+)
+
+// ScheduleILMExecution creates goroutines to manage ILM with custom policy for key/value data:
+// ILM would normally be managed inside Elasticsearch, but the peculiar nature of our data
+// is still not part of an ES ILM logic yet (see ilm.go for more details).
+//
+// Each index has a separate goroutine managing it, the shutdown via context
+// and ILM policy is shared across all goroutines.
+func ScheduleILMExecution(ctx context.Context, logger *logp.Logger, cfg config.ProfilingConfig) {
+	const (
+		maxRetries = 4
+		jitter     = 0.2
+	)
+	// create an ES client that can fulfill the ILM operations
+	client, err := es.NewClient(es.Config{
+		Addresses: cfg.ESConfig.Hosts,
+		APIKey:    base64.StdEncoding.EncodeToString([]byte(cfg.ESConfig.APIKey)),
+		// disable retries as we have dedicated retry mechanism
+		DisableRetry: true,
+	})
+	if err != nil {
+		logger.Errorf("Can't create client to perform custom ILM strategy: %v", err)
+		return
+	}
+
+	policy := ilmPolicy{
+		cfg.ILMConfig.SizeInBytes,
+		cfg.ILMConfig.Age,
+	}
+	for _, idx := range keyValueIndices {
+		go func(index string) {
+			// Jitter and the atomicity of rollover/swap give enough guarantee that
+			// we avoid collisions between multiple instances of the collector.
+			// A more robust concurrency control is implemented in the rollover function.
+			randomizedTicker := time.NewTicker(AddJitter(cfg.ILMConfig.Interval, jitter))
+			runner := &optimisticAliasSwap{
+				client:       client,
+				currentAlias: index,
+				logger:       logger,
+			}
+			for {
+				// runILM is a blocking call, waiting for the ticker's tick or context expiration.
+				if err := runILM(ctx, randomizedTicker.C, maxRetries, runner, policy); err != nil {
+					if errors.Is(err, context.Canceled) {
+						return
+					}
+					logger.Warnf("Custom ILM strategy: %v", err)
+				}
+				randomizedTicker.Reset(AddJitter(cfg.ILMConfig.Interval, jitter))
+			}
+		}(idx)
+	}
+}
+
+// ILMRunner describes the behavior of a custom ILM strategy executor
+type ILMRunner interface {
+	configure() error
+	preFlight(policy ilmPolicy) error
+	execute(policy ilmPolicy) error
+}
+
+type optimisticAliasSwap struct {
+	client *es.Client
```

Comment:
We can't rely on the client defined in other parts of apm-server, at least this is what I could grasp.
Is there a concern in adding the v8 client here?

Reply:
It's fine. I think you could use the internal/elasticsearch.Client interface by using the esapi request struct types, but it doesn't matter.